zenodo/modules/auditor/utils.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Zenodo.
+# Copyright (C) 2016 CERN.
+#
+# Zenodo is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Zenodo is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Zenodo; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Utilities for Zenodo-Auditor."""
+
+
+import logging
+from collections import Counter
+from logging.handlers import MemoryHandler
+
+from flask import current_app, url_for
+from invenio_db import db
+from invenio_pidstore.models import PersistentIdentifier, PIDStatus
+from invenio_records.models import RecordMetadata
+from mock import MagicMock
+
+from ..records.api import ZenodoRecord
+from .loggers import DBHandler
+
+
+def all_records():
+    """Return a ZenodoRecord generator with all records."""
+    records = (db.session.query(RecordMetadata)
+               .join(PersistentIdentifier,
+                     RecordMetadata.id == PersistentIdentifier.object_uuid)
+               .filter(PersistentIdentifier.pid_type == 'recid',
+                       PersistentIdentifier.status == PIDStatus.REGISTERED))
+    return (ZenodoRecord(data=r.json, model=r) for r in records)
+
+
+class tree(dict):
```

Comment:
Originally I used and preferred this for its brevity, but from [this SO answer](http://stackoverflow.com/a/19829714) I changed my mind. The `defaultdict` implementation had issues with serialization, comparisons, etc.